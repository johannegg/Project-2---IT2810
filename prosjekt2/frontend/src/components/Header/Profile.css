.profile {
    width: 180px;
    background-color: transparent;
    z-index: 10;
}

.profile-login {
    display: flex;
    flex-direction: row;
    align-items: center;
    background-color: white;
    border-radius: 5px;
    height: fit-content;
    padding-left: 2px;
    justify-content: space-between;
    right: 10px;
    min-width: 140px;
}

.login-input {
    border-color: transparent;
    width: 100%;
    outline: none;
}

.login-button {
    background-color: #E27396;
    border-top-right-radius: inherit;
    border-bottom-right-radius: inherit;
}
  
.dropdown {
    position: absolute;
    display: inline-block;
    right: 10px;
    bottom: 10px;
}

.dropdown-content {
    display: none;
    position: absolute;
    background-color: #f9f9f9;
    min-width: 140px;
    z-index: 1;
    right: 5px;
    cursor: pointer;
    border-radius: 5px;
  }
  
  .dropdown-content a {
    color: black;
    font-size: smaller;
    padding: 10px 10px;
    display: flex;
    flex-direction: column;
    border-radius: 5px;
  }
  
.dropdown-content a:hover {background-color: #f1f1f1}

.dropdown:hover .dropdown-content {
    display: block;
}

  