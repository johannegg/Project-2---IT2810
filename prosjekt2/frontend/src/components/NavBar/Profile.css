.profile-login {
    display: flex;
    flex-direction: row;
    align-items: center;
    background-color: white;
    width: 15%;
    border-radius: 5px;
    height: fit-content;
    padding-left: 2px;
    justify-content: space-between;
    position: absolute;
    right: 10px;
}

.login-input {
    border-color: transparent;
    width: 100%;
    outline: none;
}

.profile {
    background-color: transparent;
    z-index: 10;
}

.login-button {
    background-color: #E27396;
    border-top-right-radius: inherit;
    border-bottom-right-radius: inherit;
}
  
.dropdown {
    position: absolute;
    display: inline-block;
    right: 10px;
    bottom: 10px;
}

.dropdown-content {
    display: none;
    position: absolute;
    background-color: #f9f9f9;
    min-width: 140px;
    z-index: 1;
    right: 5px;
    cursor: pointer;
    border-radius: 5px;
  }
  
  .dropdown-content a {
    color: black;
    font-size: smaller;
    padding: 10px 10px;
    display: flex;
    flex-direction: column;
    border-radius: 5px;
  }
  
.dropdown-content a:hover {background-color: #f1f1f1}

.dropdown:hover .dropdown-content {
    display: block;
}

  