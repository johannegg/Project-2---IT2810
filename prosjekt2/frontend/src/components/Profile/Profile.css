.login-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    background: rgba(0, 0, 0, 0.5);
    z-index: 1005;
}

.profile-login {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    background-color: white;
    width: 300px;
    min-height: 140px;
    border-radius: 5px;
    padding: 8px;
    box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
}

.login-information {
    padding-left: 4px;
    padding-right: 4px;
}

.login-form {
    display: flex;
    flex-direction: row;
    width: 100%;
    margin-bottom: 10px;
}

.login-input {
    border: 1px solid #ccc;
    width: 100%;
    outline: none;
    padding: 5px;
    margin-right: 5px;
    background-color: white;
}

.profile-icon {
    background-color: transparent;
    z-index: 1006;
    margin: 20px;
}

.login-button {
    background-color: #E27396;
    border-top-right-radius: inherit;
    border-bottom-right-radius: inherit;
    padding: 5px 10px;
}

.dropdown {
    display: inline-flex;
    flex-direction: column;
}

.dropdown-content {
    display: none;
    position: absolute;
    top: 70%;
    min-width: 160px;
    z-index: 1;
    right: 0;
    cursor: pointer;
    border-radius: 5px;
    background-color: var(--theme-lighter-green);

}

.dropdown-content a {
    color: black;
    font-size: smaller;
    padding: 10px;
    display: flex;
    flex-direction: column;
    border-radius: 5px;
    background-color: var(--theme-lighter-green);
}

.dropdown-content a:hover {
    background-color: #f1f1f1;
    color: var(--theme-dark-pink);
}

.dropdown:hover .dropdown-content {
    display: block;
}

.profile-login-icon {
    margin: 5px;
}

.profileName {
    font-size: 13px;
    margin: 3px;
}

.close-login-button {
    position: relative;
    left: 18px;
    bottom: 66px;
    color: var(--theme-dark-grey);
}

.close-login-button:hover {
	color: #454444d8;
}

@media (max-width: 900px) {
	.profile-icon {
		font-size: medium;
		margin: 5px;
	}
}

@media (max-width: 750px) {
	.profile-icon {
		margin-left: 10px;
		margin-right: 5px;
	}
}


@media (max-width: 420px) {
	.profile-icon {
		margin: 2px;
	}
}

@media (prefers-color-scheme: dark) {
	.header {
		background-color: var(--theme-even-darker-green);
	}

	.header-title {
		color: white;
	}

	.navButton:hover {
		border-color: var(--theme-dark-pink);
	}
}



@media (prefers-color-scheme: dark) {
	.login-form {
        background-color: black;
    }

    .login-input {
        background-color: black;
        
    }

    .profile-login {
        background-color: black;
    }

    .close-login-button {
		color: #555;
	}

	.close-login-button:hover {
		color: var(--theme-lighter-green);
	}

    .login-overlay {
        background-color: rgba(28, 28, 28, 0.8);;
    }
}