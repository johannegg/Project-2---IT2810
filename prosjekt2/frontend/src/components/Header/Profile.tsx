import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowRight, faCircleUser } from '@fortawesome/free-solid-svg-icons';
import { useEffect, useState } from 'react';
import "./Profile.css";

const Profile: React.FC = () => {
    const [profileName, setProfile] = useState<string>("");
    const [isLoggedIn, setLogin] = useState<boolean>(false);
    const [inputValue, setInputValue] = useState<string>("");
    const [showLogin, setShowLogin] = useState<boolean>(false);

    const logOut = () => {
        setLogin(false);
        localStorage.removeItem("profileName");
        setShowLogin(false)
    };

    const logIn = () => {
        if (inputValue.trim() !== "") {
            setProfile(inputValue);
            setLogin(true);
            localStorage.setItem("profileName", inputValue);
        }
    };
    
    useEffect(() => {
        const storedProfileName = localStorage.getItem("profileName") || "";
        if (storedProfileName) {
            setProfile(storedProfileName);
            setLogin(true);
        }
    }, []);

    
    return (
        <>
            {isLoggedIn ? (
                <div className="dropdown">
                    <button className="profile-icon">
                        <FontAwesomeIcon icon={faCircleUser} size="2xl"/>
                    </button>
                    <div className="dropdown-content">
                        <a onClick={logOut}>
                            Log out
                        </a>
                    </div>
                </div>
            ) : (
                <>
                    <div className="dropdown">
                        <button className="profile-icon" onClick={() => setShowLogin(!showLogin)}>
                            <FontAwesomeIcon icon={faCircleUser} size="2xl"/>
                        </button>
                    </div>
                    {showLogin &&
                        <div className="login-overlay">
                        <form className="profile-login" onSubmit={(e) => { e.preventDefault(); logIn(); }}>  
                            <FontAwesomeIcon icon={faCircleUser} size="lg" style={{ color: "#E27396" }}/>
                            <input
                                className="login-input"
                                placeholder="Enter profile name"
                                onChange={(e) => setInputValue(e.target.value)}
                                />
                            <button className="login-button" type="submit">
                                <FontAwesomeIcon icon={faArrowRight} style={{ color: "#FFF" }}/>
                            </button>
                        </form>
                    </div>
                    }
                </>
            )}
        </>
    );
};

export default Profile;
