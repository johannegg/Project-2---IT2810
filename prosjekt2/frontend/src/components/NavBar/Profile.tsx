import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowRight, faCircleUser } from '@fortawesome/free-solid-svg-icons';
import { useEffect, useState } from 'react';
import "./Profile.css"

const Profile: React.FC = () => {
    const [profileName, setProfile] = useState<string>("");
    const [isLoggedIn, setLogin] = useState<boolean>(false);
    const [inputValue, setInputValue] = useState<string>("")

    const logOut = () => {
        setLogin(false)
        localStorage.removeItem("profileName");
    }

    const logIn = () => {
        if (inputValue.trim() !== "") {
            setProfile(inputValue)
            setLogin(true);
            localStorage.setItem("profileName", profileName);
        }
    };
    
    useEffect(() => {
        const storedProfileName = localStorage.getItem("profileName") || "";
        if  (storedProfileName) {
            setProfile(storedProfileName);
            setLogin(true);
        }
    }, [])


    return (
        isLoggedIn ? (
            <div className="dropdown">
                <button className="profile">
                    <FontAwesomeIcon icon={faCircleUser} size="2xl"/>
                </button>
                <div className="dropdown-content">
                    <a onClick={() => logOut()}>Log out</a>
                </div>
            </div>
        ) : (
            <form className="profile-login">    
                <FontAwesomeIcon icon={faCircleUser} size="lg" style={{color: "#E27396"}}/>
                <input
                    className="login-input"
                    placeholder="Enter profile name"
				    onChange={(e) => setInputValue(e.target.value)}
			    />
                <button className="login-button" onClick={() => logIn()}>
                    <FontAwesomeIcon icon={faArrowRight} style={{color: "#FFF"}}/>
                </button>
            </form>
        )    
    )
}

export default Profile