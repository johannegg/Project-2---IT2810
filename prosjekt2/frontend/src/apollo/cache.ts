import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
