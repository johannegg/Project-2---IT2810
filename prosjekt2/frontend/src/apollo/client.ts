import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
