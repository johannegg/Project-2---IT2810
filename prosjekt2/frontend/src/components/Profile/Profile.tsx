import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowRight, faCircleUser, faXmark } from "@fortawesome/free-solid-svg-icons";
import { useEffect, useState } from "react";
import { useMutation } from "@apollo/client";
import { CREATE_USER } from "../../utils/Queries";
import { useNavigate } from "react-router-dom";
import { favoriteSongsVar, playlistsVar } from "../../apollo/cache";
import "./Profile.css";

const Profile: React.FC = () => {
	const [profileName, setProfile] = useState<string>("");
	const [isLoggedIn, setLogin] = useState<boolean>(false);
	const [inputValue, setInputValue] = useState<string>("");
	const [showLogin, setShowLogin] = useState<boolean>(false);
	const [usernameTaken, setUsernameTaken] = useState<boolean>(false);
	
	// Initialize the createUser mutation
	const [createUser, { loading: creatingUser, error: userError }] = useMutation(CREATE_USER);
	
	const navigate = useNavigate()

	useEffect(() => {
		const storedProfileName = localStorage.getItem("profileName");
		if (storedProfileName) {
			setProfile(storedProfileName);
			setLogin(true);
		}
	}, []);

	const logOut = () => {
		setLogin(false);
		localStorage.removeItem("profileName");
		localStorage.removeItem("favoriteSongs")
		localStorage.removeItem("playlists");
		playlistsVar([]);
		favoriteSongsVar([]);
		setShowLogin(false);
		navigate("/");
	};

	const logIn = async () => {
		if (inputValue.trim() !== "") {
			try {
				const { data } = await createUser({ variables: { username: inputValue } });
				if (data && data.createUser) {
					// If user creation is successful
					setProfile(data.createUser.username);
					setUsernameTaken(false);
					setLogin(true);
					favoriteSongsVar(data.createUser.favoriteSongs);
					localStorage.setItem("profileName", data.createUser.username);
					localStorage.setItem("favoriteSongs", JSON.stringify(data.createUser.favoriteSongs));
				}
			} catch (error) {
				console.error("Error creating or retriving user:", error);
			}
		}
	};

	return (
		<>
			{isLoggedIn ? (
				<div className="dropdown">
					<button className="profile-icon">
						<FontAwesomeIcon icon={faCircleUser} size="2xl" />
					</button>
					<div className="dropdown-content">
						<p className="profileName">You're logged in as "{profileName}"</p>
						<a onClick={logOut}>Log out</a>
					</div>
				</div>
			) : (
				<>
					<div className="dropdown">
						<button className="profile-icon" onClick={() => setShowLogin(!showLogin)}>
							<FontAwesomeIcon icon={faCircleUser} size="2xl" />
						</button>
					</div>
					{showLogin && (
						<div className="login-overlay">
						<FontAwesomeIcon
							className="close-login-button"
							icon={faXmark}
							onClick={() => setShowLogin(false)}
						/>
						<div className="profile-login">
							{usernameTaken ? (
								<>
									<p>This username is already taken. Do you want to proceed?</p>
									<div className="modal-buttons">
										<button
											className="cancel-button"
											onClick={() => {
												setUsernameTaken(false);
												setInputValue(""); // Reset input value for a new attempt
											}}
										>
											No
										</button>
										<button
											className="confirm-button"
											onClick={() => {
												setShowLogin(false);
												setUsernameTaken(false);
											}}
										>
											Yes
										</button>
									</div>
								</>
							) : (
								<>
									<p className="login-information">
										{creatingUser
											? "Logging in..."
											: "Choose a unique username, or log in with an existing one"}
									</p>
									{userError && (
										<p className="error-message">{userError.message}</p>
									)}
									<form
										className="login-form"
										onSubmit={(e) => {
											e.preventDefault();
											logIn();
										}}
									>
										<FontAwesomeIcon
											className="profile-login-icon"
											icon={faCircleUser}
											size="lg"
											style={{ color: "#ea9ab2" }}
										/>
										<input
											className="login-input"
											placeholder="Enter username"
											value={inputValue}
											onChange={(e) => setInputValue(e.target.value)}
										/>
										<button className="login-button" type="submit">
											<FontAwesomeIcon icon={faArrowRight} style={{ color: "#FFF" }} />
										</button>
									</form>
								</>
							)}
						</div>
					</div>
					)}
				</>
			)}
		</>
	);
};

export default Profile;


