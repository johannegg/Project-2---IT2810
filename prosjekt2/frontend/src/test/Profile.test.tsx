import React from "react";
import { describe, test, vi, beforeEach, expect } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import Profile from "../components/Profile/Profile";
import { favoriteSongsVar, playlistsVar } from "../apollo/cache";
import { CREATE_USER } from "../utils/Queries";

// Mock `useNavigate` from react-router-dom
vi.mock("react-router-dom", () => ({
    useNavigate: () => vi.fn(), 
}));

// Test suite for the Profile component
describe("Profile Component", () => {
    const mockUsername = "testUser";

    // Mock response for the CREATE_USER mutation
    const mockCreateUserResult = {
        request: {
            query: CREATE_USER,
            variables: { username: mockUsername }, 
        },
        result: {
            data: {
                createUser: {
                    username: mockUsername, 
                    favoriteSongs: [], 
                },
            },
        },
    };

    // Before each test, clear local storage and reset all mocks
    beforeEach(() => {
        localStorage.clear();
        vi.resetAllMocks(); 
    });

    // Snapshot Tests
    test("matches snapshot when user is not logged in", () => {
        const { container } = render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );
        expect(container).toMatchSnapshot();
    });

    test("matches snapshot when login overlay is displayed", () => {
        const { container } = render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );

        // Simulate clicking on the profile icon to open the login overlay
        const profileIcon = screen.getByRole("button", { name: /profile icon/i });
        fireEvent.click(profileIcon);

        expect(container).toMatchSnapshot();
    });

    test("matches snapshot when user is logged in", async () => {
        localStorage.setItem("profileName", mockUsername);

        const { container } = render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );

        // Wait for the logged-in state to update
        await waitFor(() => {
            expect(screen.getByText(`You're logged in as "${mockUsername}"`)).toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });

    // Test to verify that the Profile component renders correctly when the user is not logged in
    test("renders login state by default", () => {
        render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );

        // Check if the profile icon is rendered
        const profileIcon = screen.getByRole("button", { name: /profile icon/i });
        expect(profileIcon).toBeInTheDocument();

        // Verify that the logged-in message is not displayed
        expect(screen.queryByText(/you're logged in as/i)).not.toBeInTheDocument();
    });

    // Test to check that the login overlay is displayed when the profile icon is clicked
    test("displays login overlay when profile icon is clicked", () => {
        render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );

        // Simulate clicking on the profile icon
        const profileIcon = screen.getByRole("button", { name: /profile icon/i });
        fireEvent.click(profileIcon);

        // Verify that the login overlay with the login prompt is displayed
        expect(screen.getByText(/choose a unique username/i)).toBeInTheDocument();
    });

    // Test to ensure a user can log in successfully
    test("logs in a user successfully", async () => {
        render(
            <MockedProvider mocks={[mockCreateUserResult]} addTypename={false}>
                <Profile />
            </MockedProvider>
        );

        // Simulate opening the login overlay
        const profileIcon = screen.getByRole("button", { name: /profile icon/i });
        fireEvent.click(profileIcon);

        // Input the mock username
        const usernameInput = screen.getByPlaceholderText(/enter username/i);
        fireEvent.change(usernameInput, { target: { value: mockUsername } });

        // Submit the login form
        const submitButton = screen.getByRole("button", { name: /submit login/i });
        fireEvent.click(submitButton);

        // Wait for the logged-in state to update
        await waitFor(() => {
            expect(screen.getByText(`You're logged in as "${mockUsername}"`)).toBeInTheDocument();
        });

        // Verify that the username is stored in local storage
        expect(localStorage.getItem("profileName")).toBe(mockUsername);

        // Check that the Apollo reactive variables are updated
        expect(favoriteSongsVar()).toEqual([]);
    });

    // Test to verify that an error message is displayed when login fails
    test("displays an error if login fails", async () => {
        const mockErrorResult = {
            request: {
                query: CREATE_USER,
                variables: { username: mockUsername },
            },
            error: new Error("User creation failed"), 
        };

        render(
            <MockedProvider mocks={[mockErrorResult]} addTypename={false}>
                <Profile />
            </MockedProvider>
        );

        // Simulate opening the login overlay
        const profileIcon = screen.getByRole("button", { name: /profile icon/i });
        fireEvent.click(profileIcon);

        // Input the mock username
        const usernameInput = screen.getByPlaceholderText(/enter username/i);
        fireEvent.change(usernameInput, { target: { value: mockUsername } });

        // Submit the login form
        const submitButton = screen.getByRole("button", { name: /submit login/i });
        fireEvent.click(submitButton);

        // Wait for the error message to appear
        await waitFor(() => {
            expect(screen.getByText(/user creation failed/i)).toBeInTheDocument();
        });
    });

    // Test to ensure a user can log out successfully
    test("logs out the user successfully", () => {
        localStorage.setItem("profileName", mockUsername);

        render(
            <MockedProvider>
                <Profile />
            </MockedProvider>
        );

        // Verify the logged-in message is displayed
        expect(screen.getByText(`You're logged in as "${mockUsername}"`)).toBeInTheDocument();

        // Simulate clicking the log-out button
        const logoutButton = screen.getByRole("button", { name: /log out/i });
        fireEvent.click(logoutButton);

        // Verify the logged-in message is no longer displayed
        expect(screen.queryByText(`You're logged in as "${mockUsername}"`)).not.toBeInTheDocument();

        // Check that local storage is cleared
        expect(localStorage.getItem("profileName")).toBeNull();

        // Verify that the Apollo reactive variables are reset
        expect(favoriteSongsVar()).toEqual([]);
        expect(playlistsVar()).toEqual([]);
    });
});