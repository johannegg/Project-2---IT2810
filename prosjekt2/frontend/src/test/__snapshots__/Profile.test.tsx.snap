// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile Component > matches snapshot when login overlay is displayed 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="true"
      aria-label="Profile icon"
      class="profile-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-user fa-2xl "
        data-icon="circle-user"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
  <div
    class="login-overlay"
  >
    <svg
      aria-hidden="true"
      aria-label="Close login form"
      class="svg-inline--fa fa-xmark close-login-button"
      data-icon="xmark"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
        fill="currentColor"
      />
    </svg>
    <div
      aria-labelledby="login-form-title"
      class="profile-login"
    >
      <p
        class="login-information"
        id="login-form-title"
      >
        Choose a unique username, or log in with an existing one
      </p>
      <form
        aria-label="Login form"
        class="login-form"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-user fa-lg profile-login-icon"
          data-icon="circle-user"
          data-prefix="fas"
          focusable="false"
          role="img"
          style="color: rgb(234, 154, 178);"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
            fill="currentColor"
          />
        </svg>
        <input
          aria-label="Enter username"
          class="login-input"
          placeholder="Enter username"
          value=""
        />
        <button
          aria-label="Submit login"
          class="login-button"
          type="submit"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-arrow-right "
            data-icon="arrow-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="color: rgb(255, 255, 255);"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
              fill="currentColor"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Profile Component > matches snapshot when user is logged in 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="true"
      aria-label="Profile icon menu"
      class="profile-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-user fa-2xl "
        data-icon="circle-user"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="dropdown-content"
    >
      <p
        aria-live="polite"
        class="profileName"
      >
        You're logged in as "
        testUser
        "
      </p>
      <a
        aria-label="Log out"
        role="button"
      >
        Log out
      </a>
    </div>
  </div>
</div>
`;

exports[`Profile Component > matches snapshot when user is not logged in 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-label="Profile icon"
      class="profile-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-user fa-2xl "
        data-icon="circle-user"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M399 384.2C376.9 345.8 335.4 320 288 320l-64 0c-47.4 0-88.9 25.8-111 64.2c35.2 39.2 86.2 63.8 143 63.8s107.8-24.7 143-63.8zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zm256 16a72 72 0 1 0 0-144 72 72 0 1 0 0 144z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
